### ISSUE TITLE
Support for multiple worlds

### ISSUE THREAD

**Issue by osrf-migration** [gazebosim/gz-sim#18]
**Original report ([archived issue](https://osrf-migration.github.io/ignition-gh-pages/#!/ignitionrobotics/ign-gazebo/issues/18)) by Louise Poubel (Bitbucket: [chapulina](https://bitbucket.org/%7B5cfa2075-477b-4ded-bdb9-8d2479544ec4%7D/), GitHub: [chapulina](https://github.com/chapulina)).**

----------------------------------------
# Description 

Since the very beginning we've had support for multiple `<world>` tags within a single SDF file. The current behaviour is that a simulation runner is started for each of them. However, as the code evolved, that use case became less well supported.

# Steps to Reproduce

At the moment, the following crashes with an ODE assertion:

`ign-gazebo -f test/worlds/multiple_worlds.sdf`

~~~
ODE INTERNAL ERROR 1: assertion "!colliders_initialized" failed in dInitColliders() [collision_kernel.cpp:168]

ODE INTERNAL ERROR 1: assertion "!colliders_initialized" failed in dInitColliders() [collision_kernel.cpp:168]
~~~

**Expected behavior:**  

At a minimum, no crash. At a maximum, multiple worlds spun up in parallel and running in lockstep, the UI displaying all of them side-by-side.

**Actual behavior:**

Server crashes, UI freezes.

**Reproduces how often:** 

100% of the time

# Versions

ign-gazebo1

# Additional Information

I'm not sure if it's best at the moment to:

* print a warning and only load the first world
* catch the physics problem before the crash and only disable the bit that is having trouble running in the same process (some singleton within ODE?)
* leave it as it is but discourage people from trying to use it. It should be possible to avoid the crash, like using a different physics engine or even no physics.

**Comment by traversaro** [comment]
> **Original comment by Addisu Z. Taddese (Bitbucket: [azeey](https://bitbucket.org/%7B9d7fdc68-d270-4080-bf03-4df0d0dbf635%7D/), GitHub: [azeey](https://github.com/azeey)).**
> 
> Looks like there is a [global cache](https://bitbucket.org/odedevs/ode/src/a77b1d06e3a62b0116e7d3c4be4c6beaac8726ea/ode/src/collision_trimesh_internal.h#lines-203) used by the collision detector in ODE. One thing to try would be to build ode with `--enable-ou`,

We recently experienced this again with @xela-95 in https://github.com/robotology/gz-sim-yarp-plugins/pull/153 . A workaround is to change DART's collision detector not to use `ODE` . At least on conda-forge, it is relatively easy to set `ENABLE_OU` option in CMake, I checked and it does not affect the public ABI of ode. However, I am not sure if this is helpful in the case two servers are stepped by the same thread (unless a new thread is always created by the server, in that case indeed that would solve the problem).

**Comment by traversaro** [comment]
One thing that we could think of at some point in the future (clearly not in an already released version of Gazebo) is to switch the default collision detector in DART from ODE to something else, I am not sure if ODE was chosen just for historical reasons.

**Comment by azeey** [comment]
>  I am not sure if ODE was chosen just for historical reasons or indeed the is the best performing collision detector for Gazebo's needs.

If I remember correctly, it was chosen because it worked much better than `DART` or `FCL` as collision detectors. I remember `FCL` converted primitive shapes to meshes, which caused unexpected behaviors. And I think `DART` as a collision engine is missing some features. It's been a while though, so it might be worth revisiting. Also, `bullet` might be a viable option, but upstream dartsim needs some changes to make mesh-to-mesh collisions work.

**Comment by traversaro** [comment]
> remember `FCL` converted primitive shapes to meshes, which caused unexpected behaviors.

That seems to be still the case, see the corresponding DART issue: https://github.com/dartsim/dart/issues/19 (that seems to be blocked by something, but I am not sure which FCL issue, perhaps https://github.com/flexible-collision-library/fcl/issues/106 ? ).

**Comment by traversaro** [comment]
Yes, indeed looking at bit at issues it seems that each collision detector has some issues, so there is no clearly better DART collision detector, xref: https://github.com/gazebosim/gz-sim/issues/1306 .

**Comment by traversaro** [comment]
I investigate a bit the problem as it occurs in https://github.com/robotology/gz-sim-yarp-plugins/issues/201 for some specific test. I was able to fix the crash with error:

~~~
ODE INTERNAL ERROR 1: assertion "!colliders_initialized" failed in dInitColliders() [collision_kernel.cpp:168]
~~~

with the following DART patch:

~~~diff
diff --git a/dart/collision/ode/OdeCollisionDetector.cpp b/dart/collision/ode/OdeCollisionDetector.cpp
index abfd232..ca45ccc 100644
--- a/dart/collision/ode/OdeCollisionDetector.cpp
+++ b/dart/collision/ode/OdeCollisionDetector.cpp
@@ -49,6 +49,8 @@
 #include "dart/dynamics/SoftMeshShape.hpp"
 #include "dart/dynamics/SphereShape.hpp"

+#include <iostream>
+
 namespace dart {
 namespace collision {

@@ -216,9 +218,14 @@ double OdeCollisionDetector::distance(
 OdeCollisionDetector::OdeCollisionDetector()
 {
   // Initialize ODE. dInitODE is deprecated.
-  const auto initialized = dInitODE2(0);
-  assert(initialized);
-  DART_UNUSED(initialized);
+  static std::mutex odeInitMutex;
+
+  {
+    std::unique_lock<std::mutex> lock(odeInitMutex);
+    const auto initialized = dInitODE2(0);
+    assert(initialized);
+    DART_UNUSED(initialized);
+  }

   dAllocateODEDataForThread(dAllocateMaskAll);
~~~

that is probably similar to the gz-sim solution proposed by @azeey in https://github.com/gazebosim/gz-sim/commit/914d44a96ac3e097d8d5c9f6ecb92a22b203303e . This solves the test failures, but those tests use simple models, while apparently this solution still have problems for complex model (like humanoid robots). Probably we can try to extend https://github.com/robotology/gz-sim-yarp-plugins/issues/201 to also include more complex models.

**Comment by traversaro** [comment]
I noticed something unfortunate related to:

~~~
ODE INTERNAL ERROR 1: assertion "!colliders_initialized" failed in dInitColliders() [collision_kernel.cpp:168]
~~~

flaky errors. I tried to work around it  by setting another collision manager, but sometimes the test continue to fail. I then realized that the problem is that even if you set another collider, for each world an instance of `OdeCollisionDetector` is anyway created in `ConstructEmptyWorld` (see https://github.com/gazebosim/gz-physics/blob/a068d3f730b8eb2886eff72a8828930489789bc9/dartsim/src/EntityManagementFeatures.cc#L728), that is called by `ConstructSDFWorld` (https://github.com/gazebosim/gz-physics/blob/a068d3f730b8eb2886eff72a8828930489789bc9/dartsim/src/SDFFeatures.cc#L492). The collision detector is then overwritten in https://github.com/gazebosim/gz-physics/blob/a068d3f730b8eb2886eff72a8828930489789bc9/dartsim/src/WorldFeatures.cc#L41-L74, but this is too late to avoid that `dInitODE2` is called two times in two different threads, resulting in the race condition that can result in:

~~~
ODE INTERNAL ERROR 1: assertion "!colliders_initialized" failed in dInitColliders() [collision_kernel.cpp:168]
~~~
